src/app/data-model-provider/ActionReturnStatus.cpp

```diff
@@ -0,0 +1,196 @@
+#include "protocols/interaction_model/StatusCode.h"
+#include <app/data-model-provider/ActionReturnStatus.h>
+
+#include <lib/support/StringBuilder.h>
+
+namespace chip {
+namespace app {
+namespace DataModel {
+
+using Protocols::InteractionModel::ClusterStatusCode;
+using Protocols::InteractionModel::Status;
+
+namespace {
+
+bool StatusIsTheSameAsError(const ClusterStatusCode & status, const CHIP_ERROR & err)
+{
+    auto cluster_code = status.GetClusterSpecificCode();
+    if (!cluster_code.HasValue())
+    {
+        return err == CHIP_ERROR_IM_GLOBAL_STATUS_VALUE(status.GetStatus());
+    }
+
+    if (status.GetStatus() != Status::Failure)
+    {
+        return false;
+    }
+
+    return err == CHIP_ERROR_IM_CLUSTER_STATUS_VALUE(cluster_code.Value());
+}
+
+} // namespace
+
+bool ActionReturnStatus::operator==(const ActionReturnStatus & other) const
+{
+    if (mReturnStatus == other.mReturnStatus)
+    {
+        return true;
+    }
+
+    const ClusterStatusCode * thisStatus  = std::get_if<ClusterStatusCode>(&mReturnStatus);
+    const ClusterStatusCode * otherStatus = std::get_if<ClusterStatusCode>(&other.mReturnStatus);
+
+    const CHIP_ERROR * thisErr  = std::get_if<CHIP_ERROR>(&mReturnStatus);
+    const CHIP_ERROR * otherErr = std::get_if<CHIP_ERROR>(&other.mReturnStatus);
+
+    if (thisStatus && otherErr)
+    {
+        return StatusIsTheSameAsError(*thisStatus, *otherErr);
+    }
+
+    if (otherStatus && thisErr)
+    {
+        return StatusIsTheSameAsError(*otherStatus, *thisErr);
+    }
+
+    return false;
+}
+
+CHIP_ERROR ActionReturnStatus::GetUnderlyingError() const
+{
+
+    if (const CHIP_ERROR * err = std::get_if<CHIP_ERROR>(&mReturnStatus))
+    {
+        return *err;
+    }
+
+    if (const ClusterStatusCode * status = std::get_if<ClusterStatusCode>(&mReturnStatus))
+    {
+        if (status->IsSuccess())
+        {
+            return CHIP_NO_ERROR;
+        }
+
+        chip::Optional<ClusterStatus> code = status->GetClusterSpecificCode();
+
+        return code.HasValue() ? CHIP_ERROR_IM_CLUSTER_STATUS_VALUE(code.Value())
+                               : CHIP_ERROR_IM_GLOBAL_STATUS_VALUE(status->GetStatus());
+    }
+
+    chipDie();
+}
+
+ClusterStatusCode ActionReturnStatus::GetStatusCode() const
+{
+    if (const ClusterStatusCode * status = std::get_if<ClusterStatusCode>(&mReturnStatus))
+    {
+        return *status;
+    }
+
+    if (const CHIP_ERROR * err = std::get_if<CHIP_ERROR>(&mReturnStatus))
+    {
+        if (err->IsPart(ChipError::SdkPart::kIMClusterStatus))
+        {
+            return ClusterStatusCode::ClusterSpecificFailure(err->GetSdkCode());
+        }
+
+        if (*err == CHIP_NO_ERROR)
+        {
+            return ClusterStatusCode(Status::Success);
+        }
+
+        if (err->IsPart(ChipError::SdkPart::kIMGlobalStatus))
+        {
+            return ClusterStatusCode(static_cast<Status>(err->GetSdkCode()));
+        }
+
+        return ClusterStatusCode(Status::Failure);
```

Comment:
I agree. I just did not want to pull StatusIB since the order of creation seems to be `StatusCode` returned and `StatusIB created out of it` and I am replacing StatusCode

Reply:
Would it not be better to get a ClusterStatus from CHIP_ERROR instead? that fixes my code and allows for re-use. StatusIB is way down the chain (people don't return a statusib, they return a status or clusterstatus.

Reply:
I think we are in violent agreement here.  The code should live in ClusterStatusCode, and we should go through that to init a StatusIB from CHIP_ERROR.